src/manager.S

```diff
@@ -129,3 +129,67 @@ add_compartment:
     b.gt      abort
 
     ret
+
+/**
+ * Function to delete an existing compartment data
+ *
+ * @param x0 Function of compartment to be deleted
+ */
+/* TODO there is a potential issue of a compartment saving a capability it was
+ * provided to it for use (i.e., PCC, DDC), such as by storing it in the memory
+ * of another compartment. If the original compartment gets deleted, and a new
+ * compartment is allocated in the same memory region, we have leaked a
+ * capability referencing the new compartment. */
+.type del_compartment, "function"
+del_compartment:
+    // Find capabilities (DDC/PCC) of given compartment
+    ldr       x1, comps_addr
+    mov       w2, #0
+    ldr       w3, comps_cnt
+
+pcc_check_start:
+    ldr       c4, [x1, #COMP_OFFSET_PCC] // get current stored PCC to check
+    gcvalue   x4, c4
+    cmp       x4, x0 // check value of current PCC against given function
+    b.eq      pcc_found
+    add       w2, w2, #1
+    cmp       w2, w3 // compartment bounds check
+    b.eq      pcc_not_found
+    add       x1, x1, #COMP_SIZE // go check next PCC
+    b         pcc_check_start
+
+pcc_not_found:
+    b         abort
+
+pcc_found:
+    // We found the address to the start of the compartment info area, at
+    // address `$x1`
+    stp       x1, lr, [sp, #-16]! // store `x1` and `lr`
+    ldr       c2, [x1, #COMP_OFFSET_DDC]
+
+    // Unmap memory region allocated
+    gcvalue   x0, c2
+    gclen     x1, c2
+    bl        munmap
+    cmp       x0, #-1
+    b.eq      abort // munmap failed
+    ldp       x0, lr, [sp], #16
+
+    // Shift subsequent capabilities forward (if we delete the last
+    // compartment, we put gibberish instead)
```

Comment:
A bit hasty, running out of time: I also wonder if it's best to just fill these with `nullptr` to avoid the reuse issue that Laurie mentioned, but that may be the same thing that's mentioned in the TODO at the top.

Reply:
I'm unsure what you're proposing is rather similar to what is implemented. On adding a compartment, we zero out memory in the to be allocated region, via `MAP_ANONYMOUS` (`manager.S:45`). We could do this on compartment deallocation instead, but is there much difference between the two approaches?

Reply:
I wondered if there could be a lingering capability to the list entry. Sorry, I should have checked more thoroughly before posting. Please disregard :-)

Reply:
Just to clarify for myself, the TODO and what you're saying are two distinct issues. The todo has to do with a compartment saving capabilities referring to a region of memory, which will eventually get reused for another, future compartment, thus giving the original compartment access to the new compartment's memory.

What you're talking about here is whether in the switcher, there are potential left-over capabilities that might be somehow leaked. I think it *might* be possible if we are in a situation that we have all compartments allowed allocated (since we only allow a static limit of compartments), and then deleting the last allocated compartment shifts some memory in which is no gibberish, as expected. So it might be worth considering zero-ing out the memory region of the previously last compartment, than expecting to shift gibberish in. I hope that makes sense.